src/Event/DefaultRoleEventInterface.php

```diff
@@ -95,11 +104,18 @@ public function deleteRole($name);
    * Returns whether or not the given role exists.
    *
    * @param string $name
-   *   The name of the role for which to verify the existance.
+   *   The name of the role for which to verify the existence.
    *
    * @return bool
    *   TRUE if the role exists, FALSE otherwise.
    */
   public function hasRole($name);
 
+  /**
+   * Resets the internal static cache.
+   *
+   * Call this before dispatching the event.
+   */
+  public function reset();
```

Comment:
I think it's strange too, but this is a so called "ContainerAware" event, which is a service that is lazy loaded when first requested and then persists in the container (much like a singleton). This has the side effect that if a second round is done the previous results are still there. In normal use it would probably be rare (but not impossible) to fire the event listener twice, but this happens all the time in tests.

The `ContainerAwareEventDispatcher` ignores this, it doesn't seem have support to do something like an `initialize()` on dispatch. The [documentation](http://symfony.com/doc/current/components/event_dispatcher/container_aware_dispatcher.html) on it is also very sparse.

I was looking for an issue about it on the Symfony but there are only 4 open issues mentioning the container aware event dispatcher.

As an alternative to doing a manual reset, we can also cache the data locally per entity type and bundle.